### ISSUE TITLE
Feature eventing (native equivalent of queryRenderedFeatures and querySourceFeatures)

### ISSUE THREAD

**Issue by hallahan** [mapbox/mapbox-gl-native#352]
I've been playing with MapboxGL, and I'm trying to figure out how to handle click events to select features. Is this functionality built in?

Looking at the JavaScript in the several blog posts that use mapbox-gl-js, I am able to got the Point object from a click event. Then, I see a function `map.featuresAt`, but I am not able to get it to return any features...

Any clues? Regarding the native libraries, would this be in the Cocoa library?

**Comment by hallahan** [comment]
What is the status for a native `featuresAt` implementation? Is the ability to select a feature on the map going to be a feasible use case for native?

**Comment by incanus** [comment]
It will be eventually, but the focus for now has just been on point annotations, which exist in a runtime style layer and query based on `~50px` square padding for touch gestures. We'll eventually port over `featuresAt` and allow querying of all of the style layers such as OSM base data.

**Comment by tylrtrmbl** [comment]
I am so pumped for `featuresAt`. Thanks all!

**Comment by hallahan** [comment]
Being able to select OSM base data would be an extremely valuable feature. This is something I'd enjoy working on. Could you point me in the right direction?

**Comment by incanus** [comment]
Yes, I would check out the `featuresAt` functionality in `gl-js`. The basic idea is an R-tree that bins the features at parse time so that efficient spatial queries can be made at runtime in response to the "at" and a radius. The structures that exist now aren't indexing parsed features this way. 

Point annotations do index based on tiles made at runtime as annotations are added and removed, which is how gestural touch of points on the map that are rendered in GL was possible.

**Comment by hallahan** [comment]
It looks like you have boost's rtree in there. Are you using this in the same way as rbush in the JS implementation? Are you saying only points are added to a spatial index?

**Comment by hallahan** [comment]
I'm poking around, and the only actual rtree usage I am seeing is in `collision.hpp`. It looks like it's being used to make sure labels don't collide? 

https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/text/collision.hpp#L32

Would we want to use the same container as is being used here for map geometries? I know PostGIS uses `geos`. I've used JTS Topology Suite in Java, the lib geos was ported from. 

http://stackoverflow.com/questions/9891374/is-boost-geometry-mature-enough

What do you guys think? Clearly this is not a trivial feature.

**Comment by incanus** [comment]
Beginnings of this in https://github.com/mapbox/mapbox-gl-native/commit/139b9e7a60acadadaef80eea1cd07811e00f2cfb. Right now it's just inserting feature names (actually, `name_en` or `unknown` if not present) at tile parse time for style layers starting with `poi` (as an alternative to parsing `interactive: true` for now). 

I realize we eventually want this to be style-aware (which it could fairly easily be) as well as eliminate the need for `interactive: true` in the style (https://github.com/mapbox/mapbox-gl-js/issues/1479), but this is a start as I keep hitting needs for this and wanted to get my feet wet with `boost::geometry::rtree` a bit. 

Next up is to plumb through interactive gestures such as we do with annotations right now. For debug purposes, we could even piggyback on / add results to the existing annotation query, at least for log-debugging. This should just be a fairly simple `featureTree.query(boost::geometry::index::intersects())` in the style of what we do for annotations. 

What we should end up with is tapping the map logging nearby POIs when present (i.e. at high zooms) as a proof of concept. 

Parsing output right now: 

```
added poi-parks-scalerank2 - River View Cemetery
added poi-parks-scalerank2 - Waverly Country Club
added poi-parks-scalerank2 - Lewis and Clark College
added poi-parks-scalerank2 - Lewis and Clark College
added poi-parks-scalerank2 - Tryon Creek State Natural Area
added poi-parks-scalerank2 - Sellwood Park
added poi-parks-scalerank2 - Powers Marine Park
added poi-parks-scalerank2 - Beth Israel Cemetery
added poi-parks-scalerank2 - Sellwood Riverfront Park
added poi-parks-scalerank2 - Greenwood Hills Cemetery
feature tree for 14,2609,5864 [poi-parks-scalerank2] has 10 members
added poi-scalerank2 - Waverly Country Club
added poi-scalerank2 - Sellwood Park
added poi-scalerank2 - Milwaukie High School
added poi-scalerank2 - Milwaukie Elementary School
feature tree for 14,2610,5864 [poi-scalerank2] has 4 members
```

This currently does not noticeably affect rendering performance, even though it's adding to the R-tree from inside the workers ahead of render. It _does_ affect things when you add all layers, which is really the meat of https://github.com/mapbox/mapbox-gl-js/issues/1479 — automatically parsing all features at a much larger scale efficiently.

**Comment by incanus** [comment]
Getting close here. Current blocker is the fact that we postpone parsing of symbol layers until dependencies are loaded, and thus far I've just been adding to a tile's feature tree during `!partialParse`. We also need to possibly expose style bucket features to the `TileWorker`, as this is where and when we build up tile feature trees during tile parse. Right now our native API can only add features to a bucket, then pass the buckets as results back to the tile data, then later the painter asks the buckets to render themselves. We never introspect the features. What I may do is have the buckets build up feature trees internally, exposed via a getter, and have the tile data aggregate their contents post-parse rather than try to look at the features straight out of the geometry tiles in duplicate of their actual parsing.

**Comment by incanus** [comment]
The downside to having buckets build up their own feature trees is duplication of code in each bucket type.

**Comment by incanus** [comment]
Also of note here: what does a native `featuresAt` actually return? Unlike JavaScript, styles and features as JSON (or GeoJSON) is not necessarily useful or efficient here, and there are so many strongly-typed objects involved on the native side that exposing all of those to platform clients doesn't necessarily make sense either.

**Comment by hallahan** [comment]
For OpenMapKit, returning GeoJSON or vector data is not of particular use, since it's been simplified and lacks original OSM topology.

Feature selection that returns OSM node, way, and relation IDs baked into the vector tiles would be required for use in an OSM related application.

When selecting a feature, we need:

1. The node / way ID of the feature selected.
2. The IDs of the relations that contain the selected feature.
3. The OSM version number of the selected feature and relation(s).

With these 3 properties, we have what we will need to select appropriately from our application's OSM data model.

cc/ @dalekunce @mojodna

**Comment by incanus** [comment]
I'm going to spend some more time thinking about this later, but right now I'm thinking of providing the layer name as a string and then a small dictionary of the properties out of the vector tile. For Mapbox sources, this would include most of the OSM data.

**Comment by incanus** [comment]
For the (optional per JS `featuresAt`) feature geometries, I'm considering either GeoJSON or maybe a simple typed nested set of geo coordinates mirroring GeoJSON, allowing the platform (iOS/Android/etc.) to natively type it if desired.

**Comment by jfirebaugh** [comment]
I think the output types are going to be in flux until we get a Style API and GeoJSON source API. When those are in place, the output features should use the same types as the GeoJSON source API, with an additional `layer` property that uses the same layer types as the Style API.

**Comment by incanus** [comment]
Ok as of b1e3da9055a54b9af9ec72d7f6654205f5c31fd7 we have `featuresAt` functionality logging properly any properties from the layer parsing buckets. However, I'm not quite sure why we don't see all of the properties like we do in JS (e.g. `osm_id`) so taking a look at that. 

This needs some refactoring to push things closer down to the style source instead of in `MapContext`, but it's got all the major problems ironed out for functionality to occur. 

![anigif-1449020567](https://cloud.githubusercontent.com/assets/17722/11519896/f483d8aa-9852-11e5-85a6-b0edd325ea5d.gif)

**Comment by incanus** [comment]
- [ ] Move stuff down into `Source`
- [ ] Figure out why only some feature attributes are coming through
- [ ] Figure out why very occasionally initially-shown tiles don't have features indexed yet

**Comment by hallahan** [comment]
Question: can we select lines and polygons at this point, or is this only markers?

**Comment by friedbunny** [comment]
@hallahan: Only markers are selectable right now; polygon and polyline selection has been discussed over in #2082.

**Comment by hallahan** [comment]
@friedbunny - It's been some time since I last dove into the internals of this library, but from what I recall, there is a boost r-tree index of only labels and markers in the native library, but r-bush in JS has an index of everything.

I'm talking about selection of actual vector tile features, not just annotations. Being able to interact with the "base map" is really what we need.

If one is to edit OpenStreetMap features, I'd think it would make sense to be able to select the vector tile feature first, get the ID from the properties, and then fetch the non-simplified OSM entities from a data source for vertex editing. Also, if one were to edit the tags, we wouldn't really need the source geometries until actually submitting back to the OSM Editing API.

I'd love suggestions from anyone who can describe a path forward to manipulate OSM data with the mapbox-gl-native library. Right now, it seems like we are being blocked by this missing functionality.

**Comment by 1ec5** [comment]
> there is a boost r-tree index of only labels and markers in the native library, but r-bush in JS has an index of everything.

Would that change in #3172?

**Comment by 1ec5** [comment]
As of mapbox/mapbox-gl-js#2224, this feature is called `queryRenderedFeatures()` in GL JS.

**Comment by hallahan** [comment]
Hi Mapbox Devs! Is there any update to the status of this feature? It looks like the work @incanus did in December needs to be reimplemented to follow the current apprach of `queryRenderFeatures` in `mapbox-gl-js`?

Is there going to be a need for a C++ `grid-index`? Maybe not, since objects do not need to be serialized when transferring them to another thread?

Are there major disadvantages to using GEOS for things like spatial indices in GL Native? What if you used [geos::index::quadtree::Quadtree](http://geos.refractions.net/ro/doxygen_docs/html/classgeos_1_1index_1_1quadtree_1_1Quadtree.html#_details)? My apologies if this is a naïve question. I'm hoping to start a conversation on this functionality--it's something we really need if we are to build any application providing introspection into features within the vector tile.

**Comment by ansis** [comment]
>  Is there any update to the status of this feature?

I'm picking this up now. It's hard to tell how straightforward the `queryRenderedFeatures` port will be.

> Is there going to be a need for a C++ grid-index? Maybe not, since objects do not need to be serialized when transferring them to another thread?

I won't port this initially. I might check later if it can speed up label collision detection.

> Are there major disadvantages to using GEOS for things like spatial indices in GL Native? What if you used geos::index::quadtree::Quadtree?

I'm not familiar with GEOS. I'll stick to the boost rtree for the initial implementation just because we're already using it. We can change the index later if needed.

**Comment by hallahan** [comment]
Great to hear Ansis! It sounds like boost is a logical approach. 

[GEOS](https://trac.osgeo.org/geos/) is one of the core libs that make PostGIS PostGIS. It's the C++ port of the [Java Topology Suite](https://en.wikipedia.org/wiki/JTS_Topology_Suite). QGIS and lots of other tools use it.

Really excited to see this work happen :smile:

**Comment by ansis** [comment]
Pull request open at https://github.com/mapbox/mapbox-gl-native/pull/4656. It implements most of the internals. The input and return types are still undecided.

**Comment by jfirebaugh** [comment]
The core internals PR has landed. Next steps are to shepherd [geometry.hpp](https://github.com/mapbox/geometry.hpp/) to a usable state and replace the string-based API with true features.

`Filter` is now suitable for the public API, so we can integrate that as well.

Then write SDK bindings.

**Comment by jfirebaugh** [comment]
geometry.hpp is now in use and the core API now returns `Feature` objects rather than stringified JSON. However, the all features are returned with Null Island geometry. Next step is to implement https://github.com/mapbox/mapbox-gl-test-suite/issues/86 and return real geometries.

**Comment by jfirebaugh** [comment]
This is ready for implementation of SDK bindings.

**Comment by 1ec5** [comment]
Development of iOS and OS X SDK support for this feature is happening in #5110.

**Comment by 1ec5** [comment]
Now that #5110 has landed, the upcoming iOS SDK v3.3.0 beta and OS X SDK release will have `-[MGLMapView visibleFeaturesAtPoint:]` and `-[MGLMapView visibleFeaturesInRect:]` as equivalents to GL JS’s `queryRenderedFeatures()`. Remaining items before this issue can be closed:

* Android SDK equivalent to #5110
* Port `querySourceFeatures()` from mapbox/mapbox-gl-js#2052 (specifically mapbox/mapbox-gl-js@7b0a3811033836043f2995061ebb1c36bd9826df) for finding all features, not just the ones that happen to be rendered

**Comment by tobrun** [comment]
I have started to work on implementing Android equivalent of #5110 in (will update with PR number once that lands). Before implementing the contents of #5110, I need to refactor the Android annotations API to match iOS more. As pointed out by @tmcw in https://github.com/mapbox/mapbox-gl-native/issues/3288 and related, there are some fundamental issues with the current Android API. Matching the iOS API will not only help implementing this feature but will also make it easier for implementing parity features in the long run.

**Comment by 1ec5** [comment]
Note that the iOS SDK is affected by #5201 just like the Android SDK is affected by #3288. If you do make MultiPoint a separate concept from Polyline and Polygon, it'll be a backwards-incompatible change. One alternative I proposed in https://github.com/mapbox/mapbox-gl-native/issues/5201#issuecomment-222858385 was to create a new class to represent the GeoJSON concept of a multipoint, but call it something else.

**Comment by tobrun** [comment]
@1ec5 
Since we have seen some resistance on the name Multipoint, I would suggest naming this exact the same as the GeoJSON equivalent. 

- Do you have any suggestions for a name for the parent of Polygon, Polyline and Multipoint?  

  FYI: this is the current inheritance scheme:

  > 
Annotation > Shape > Marker
Annotation > Shape > ? > Polygon
Annotation > Shape > ? > PolyLine
Annotation > Shape > ? > Multipoint

- Multipoint is currently empty since all logic is encapsulated in `?`, do you have the same issue on iOS or am I missing something here?

**Comment by 1ec5** [comment]
The discussion in #5201 is specific to the iOS SDK, which has MGLMultiPoint for compatibility with MapKit. That class actually does have code behind it, privately, which is why it can’t just be a protocol. If you’re OK with a backwards-incompatible change in the Android SDK, I suggest making all four classes direct children of `Shape`, for consistency with GeoJSON.

**Comment by incanus** [comment]
Breaking out port of `querySourceFeatures` into https://github.com/mapbox/mapbox-gl-native/issues/5792 since it's of different priority than wrapping up the Android side of `queryRenderedFeatures` right now.

**Comment by tobrun** [comment]
@ivovandongen created an android follow up issue in #5869. 
Closing as all other items were ticketed out or resolved.
